declarations/index.d.ts

```diff
@@ -1,22 +1,107 @@
 /// <reference types="node" />
 import { EventEmitter } from 'events';
-export interface IAbstractIO {
-    isReady: boolean;
-    readonly MODES: {
-        INPUT: number;
-        OUTPUT: number;
-        ANALOG: number;
-        PWM: number;
-        SERVO: number;
-    };
+export declare enum Mode {
+    INPUT = 0,
+    OUTPUT = 1,
+    ANALOG = 2,
+    PWM = 3,
+    SERVO = 4,
+    UNKOWN = 99
+}
+export declare enum Value {
+    HIGH = 1,
+    LOW = 0
+}
+export interface IPinConfiguration {
+    supportedModes: Mode[];
+    mode: Mode;
+    value: Value;
+    reporting: 0 | 1;
+    analogChannel: number;
+}
+export interface IPingReadSettings {
+    pin: number | string;
+    value?: Value;
+    pulseOut?: number;
+}
+export interface II2CConfig {
+    address: number;
+    bus?: number | string;
+    port?: number;
+    delay?: number;
+}
+export interface ISerialConfig {
+    portId: number | string;
+    baud?: number;
+    rxPin?: number | string;
+    txPin?: number | string;
+}
+export interface IServoConfig {
+    pin: number | string;
+    min?: number;
+    max?: number;
 }
-export declare class AbstractIO extends EventEmitter implements IAbstractIO {
-    isReady: boolean;
+export declare type ReadHandler = (data: number[]) => void;
+export declare class AbstractIO extends EventEmitter {
     readonly MODES: {
-        INPUT: number;
-        OUTPUT: number;
-        ANALOG: number;
-        PWM: number;
-        SERVO: number;
+        INPUT: Mode;
+        OUTPUT: Mode;
+        ANALOG: Mode;
+        PWM: Mode;
+        SERVO: Mode;
+        UNKOWN: Mode;
     };
+    readonly HIGH: Value;
+    readonly LOW: Value;
+    readonly pins: IPinConfiguration[];
+    readonly analogPins: number[];
+    readonly name: string;
+    readonly defaultLed: string | number | undefined;
+    readonly isReady: boolean;
+    pinMode(pin: string | number, mode: Mode): void;
+    pwmWrite(pin: string | number, value: number): void;
+    servoWrite(pin: string | number, value: number): void;
+    digitalWrite(pin: string | number, value: number): void;
```

Comment:
A couple of recent changes to digitalWrite allow a user to enqueue multiple digital writes before sending a single port update. When ```enqueue === false```, the digital write is immediate. 
````
digitalWrite(pin: string | number, value: number, enqueue = false): void;
flushDigitalPorts(): void;
````
For platforms that don't write by port or support bulk digital writes in some other way, there are three ways to handle it: 
1. Always write immediately and have ```flushDigitalPorts()``` do nothing. 
1. Queue up writes and just call the platform's write method multiple times in quick succession on flush.
1. Ignore it. J5 assumes not all IO-plugins will support this so it checks before calling.

Documentation at https://github.com/firmata/firmata.js/tree/master/packages/firmata.js#pin is not correct so I'm glad this came up.

Reply:
Oh interesting. I've added an entry for it that does nothing if not overridden (as opposed to throwing an exception). Does this sound right to you?

Reply:
I think I should throw. If a user expects X but the plugin fails silently, that could be really frustrating.